Handle Expected<T> from TempFile::create() to prevent potential crashes

devajithvs opened the pull request:
Add proper error handling for `TempFile::create()` by checking `takeError()` and `discard()`

I get a crash with the following error without this change:

```
Expected<T> must be checked before access or destruction.
Expected<T> value was in success state. (Note: Expected<T> values in
success mode must still be checked prior to being destroyed).
```

illuhad reviewed (commented):
Sorry, this PR kind of went under my radar, apologies.

Do we really need the if-statements? My guess is that LLVM just checks whether the error has been cast to bool. In this case we could solve the problem without creating useless branching everywhere:

```c++
AtScopeExit DestroyInputFile([&]() {
  auto Err = InputFile->discard(); static_cast<bool>(Err);
});
```

We could make it a bit more readable using some thing like


```c++
#define ACPP_DISCARD_LLVM_ERROR(Expression) \
  {auto Err = (Expression); static_cast<bool>(Err); }

AtScopeExit DestroyInputFile([&]() {
  ACPP_DISCARD_LLVM_ERROR(InputFile->discard());
});
```

devajithvs committed:
Handle Expected<T> from TempFile::create() to prevent potential crashes

Add proper error handling for `TempFile::create()` by checking
`takeError()` for output files and ensure proper cleanup with
`discard()` and log errors

I get the following error without this change:

```
Expected<T> must be checked before access or destruction.
Expected<T> value was in success state. (Note: Expected<T> values in
success mode must still be checked prior to being destroyed).
```

devajithvs commented:
Thanks @illuhad,

I realized that LLVM already provides consumeError() for this exact purpose. I've now replaced all branching occurrences with `consumeError()`

illuhad reviewed (approved):
Lovely, thanks! Let's just wait for CI to come back, then we can merge :)
